### PR TITLE
[WIP] Arnold Standard Material

### PR DESCRIPTION
**Pull request by fire**
Don't merge yet.

Putting code for review.

Don't have an emission sample scene.

Transparency doesn't work yet.

Edited:

Crash when exporting as glb.

```
maya2glTF -sn "flashbang" -of "C:/Users/fire/Downloads/10_test_case_flashbang/10_test_case_flashbang/export/" -sf 0.01 -cpr "(C)2019"  -acn "clip" -ast 1 -aet 120 -afr 24 -sno -vno
maya2glTF: loading plugin...
maya2glTF: plugin loaded.
// Error: line 1: Image with path '' does not exist! // 
// Error: line 1: (it is advised to use a Maya project and relative paths) // 
// Error: line 1: Mesh 'flashbang_03_low_flashbang_03_lowShape' has 549 invalid tangents!
Assign texture coordinates and/or cleanup your mesh and try again please.
Use the following command to visualize the tangents and select the first invalid face-vertices:
```

IMVU sponsored.

### CODE REVIEW

File: src/ExportableMaterial.h [hunk 1]
```diff
@@ -88,6 +90,12 @@ class ExportableMaterialPBR : public ExportableMaterialBasePBR {
 
     template <class MFnShader>
     void convert(ExportableResources &resources, const MObject &shaderObject);
+    void loadAiStandard(ExportableResources &resources,
+                        const MFnDependencyNode &shaderNode);
+
+    MStatus tryCreateRoughessMetalnessTexture(ExportableResources &resources, const ExportableTexture &metallicTexture,
```
Comment (fire) [inline src/ExportableMaterial.h]: Move to be with the try create normal texture?

File: src/ExportableMaterial.cpp [hunk 2]
```diff
@@ -172,6 +178,196 @@ void ExportableMaterialPBR::convert(ExportableResources &resources,
...
+
+        m_glMetallicRoughness.metallicRoughnessTexture =
+            &m_glMetallicRoughnessTexture;
+    }
+
+    // Emissive color
+    m_glEmissiveFactor = {0, 0, 0, 0};
+    if (getColor(shaderObject, "emission", m_glEmissiveFactor)) {
+        m_glMaterial.emissiveFactor = &m_glEmissiveFactor[0];
+    }
```
Comment (fire) [inline src/ExportableMaterial.cpp]: Untested emission code.

File: src/ExportableMaterial.cpp [hunk 3]
```diff
@@ -172,6 +178,196 @@ void ExportableMaterialPBR::convert(ExportableResources &resources,
...
+            getScalar(shaderObject, "baseWeight", customColorWeight);
+        if (hasWeight) {
+            m_glMetallicRoughness.baseColorFactor = &customColorWeight;
+        }
+        m_glMaterial.metallicRoughness = &m_glMetallicRoughness;
+    }
+
+    if (customBaseColor[3] != 1.0f) {
+        m_glMaterial.alphaMode = "BLEND";
+    }
```
Comment (fire) [inline src/ExportableMaterial.cpp]: Doesn't work yet.

File: src/ExportableMaterial.cpp [hunk 4]
```diff
@@ -172,6 +178,196 @@ void ExportableMaterialPBR::convert(ExportableResources &resources,
...
+    const auto shaderObject = shaderNode.object(&status);
+    THROW_ON_FAILURE(status);
+
+    auto &args = resources.arguments();
+    args.assignName(m_glMaterial, shaderNode.name().asChar());
+
+    float opacityFactor = 1.f;
+
+    auto hasOpacity =
+        getScalar(shaderObject, "TransmissionWeight", opacityFactor);
```
Comment (fire) [inline src/ExportableMaterial.cpp]: Buggy?

Comment (Ziriax) [reply src/ExportableMaterial.cpp]: What would be buggy here? I'm not familiar with the Arnold material, so reviewing is a bit tricky without digging into the API of that material myself.

File: src/ExportableMaterial.h (hunk 1 above)
Comment (Ziriax) [reply src/ExportableMaterial.h]: > Move to be with the try create normal texture?

Not sure what you mean with this.

The code looks very good, only a little typo in 
`tryCreateRoughessMetalnessTexture`, should be 
`tryCreateRoughnessMetalnessTexture`

File: src/ExportableMaterial.cpp (hunk 4 above)
Comment (fire) [reply src/ExportableMaterial.cpp]: I can’t find the correct attribute here for opacity and the documentation is lacking. It also requires the transparency check.

**Comment by fire** [comment]
Will be focusing on https://github.com/WonderMediaProductions/Maya2glTF/issues/93 and https://github.com/WonderMediaProductions/Maya2glTF/issues/88 before this.

File: src/ExportableMaterial.cpp (hunk 4 above)
Comment (Ziriax) [reply src/ExportableMaterial.cpp]: Well, in that case, I can't help you right now with this. 

If the AI export works for you, then I'm ready to approve this code, just let me know.

**Comment by Ziriax** [comment]
I think I will already merge this, it seems like a good start. It can be improved later. My conversion from lambert to PBR is also very limited.

Okay for you?

**Comment by fire** [comment]
I'll rebase then give me a few moments.

**Comment by fire** [comment]
I'm currently on the mac so the patch is on top of the Mac pr.

Pushed, but will do some testing.

**Comment by fire** [comment]
Doing testing after the rebase, will try to export my sample models.

**Comment by Ziriax** [comment]
> Doing testing after the rebase, will try to export my sample models.

Okay, I'll merge the pull request as soon as you give me a sign

**Comment by fire** [comment]
Something changed:

Getting compile errors on.

```
maya #define USING_LATEST_API USING_MAJOR_API
```

```
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableAsset.cpp:7:
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableAsset.h:2:
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableClip.h:5:
In file included from /Users/mac/Documents/Maya2glTF/src/NodeAnimation.h:4:
In file included from /Users/mac/Documents/Maya2glTF/src/PropAnimation.h:4:
In file included from /Users/mac/Documents/Maya2glTF/src/accessors.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/MeshRenderables.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/Mesh.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/MeshShape.h:3:
/Users/mac/Documents/Maya2glTF/src/MayaException.h:16:12: error: reference to
      'MStatus' is ambiguous
    static MStatus printError(const std::string &message, const MStatus &error);
           ^
/Users/mac/Documents/Maya2glTF/src/ExportableMaterial.h:77:7: note: candidate
      found by name lookup is 'MStatus'
class MStatus;
      ^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:1300:1: note:
      candidate found by name lookup is 'MStatus'
USING_LATEST_API
^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:1291:26: note:
      expanded from macro 'USING_LATEST_API'
#define USING_LATEST_API USING_MAJOR_API
                         ^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:1289:2: note:
      expanded from macro 'USING_MAJOR_API'
        USING_MAJOR_ONLY_API
        ^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:984:44: note:
      expanded from macro 'USING_MAJOR_ONLY_API'
    using OPENMAYA_MAJOR_NAMESPACE_SCHEMA::MStatus;                            \
                                           ^
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableAsset.cpp:7:
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableAsset.h:2:
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableClip.h:5:
In file included from /Users/mac/Documents/Maya2glTF/src/NodeAnimation.h:4:
In file included from /Users/mac/Documents/Maya2glTF/src/PropAnimation.h:4:
In file included from /Users/mac/Documents/Maya2glTF/src/accessors.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/MeshRenderables.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/Mesh.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/MeshShape.h:3:
/Users/mac/Documents/Maya2glTF/src/MayaException.h:16:65: error: reference to
      'MStatus' is ambiguous
    static MStatus printError(const std::string &message, const MStatus &error);
                                                                ^
/Users/mac/Documents/Maya2glTF/src/ExportableMaterial.h:77:7: note: candidate
      found by name lookup is 'MStatus'
class MStatus;
      ^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:1300:1: note:
      candidate found by name lookup is 'MStatus'
USING_LATEST_API
^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:1291:26: note:
      expanded from macro 'USING_LATEST_API'
#define USING_LATEST_API USING_MAJOR_API
                         ^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:1289:2: note:
      expanded from macro 'USING_MAJOR_API'
        USING_MAJOR_ONLY_API
        ^
/Applications/Autodesk/maya2019/include/maya/MApiNamespace.h:984:44: note:
      expanded from macro 'USING_MAJOR_ONLY_API'
/Users/mac/Documents/Maya2glTF/src/ExportableCamera.cpp:14:5:     using OPENMAYA_MAJOR_NAMESPACE_SCHEMA::MStatus;                            \
error                                           ^:
no type
      named 'MStatus' in 'MayaException'; did you mean 'draco::Status'?
    MayaException::MStatus status;
    ^~~~~~~~~~~~~~~~~~~~~~
    draco::Status
/Users/mac/Documents/Maya2glTF/build/COLLADA2GLTF/src/COLLADA2GLTF/GLTF/dependencies/draco/src/draco/core/status.h:24:7: note:
      'draco::Status' declared here
class Status {
      ^
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableAsset.cpp:7:
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableAsset.h:2:
In file included from /Users/mac/Documents/Maya2glTF/src/ExportableClip.h:5:
In file included from /Users/mac/Documents/Maya2glTF/src/NodeAnimation.h:4:
In file included from /Users/mac/Documents/Maya2glTF/src/PropAnimation.h:4:
In file included from /Users/mac/Documents/Maya2glTF/src/accessors.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/MeshRenderables.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/Mesh.h:3:
In file included from /Users/mac/Documents/Maya2glTF/src/MeshShape.h:3:
/Users/mac/Documents/Maya2glTF/src/MayaException.h:6:33: error: variable has
      incomplete type 'const MStatus'
    MayaException(const MStatus error, const std::string &message)
                                ^
/Users/mac/Documents/Maya2glTF/src/ExportableMaterial.h:77:7: note: forward
      declaration of 'MStatus'
class MStatus;
      ^
/Users/mac/Documents/Maya2glTF/src/ExportableCamera.cpp:24:12: error: no viable
      overloaded '='
    status = sl.getDependNode(0, defaultResolutionNode);
    ~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/mac/Documents/Maya2glTF/build/COLLADA2GLTF/src/COLLADA2GLTF/GLTF/dependencies/draco/src/draco/core/status.h:50:11: note:
      candidate function not viable: no known conversion from
      'Autodesk::Maya::OpenMaya20190000::MStatus' to 'const draco::Status' for
      1st argument
  Status &operator=(const Status &) = default;
```

**Comment by fire** [comment]
It's nasty, I have to be at a meeting, so any comments for the work would be helpful.

**Comment by fire** [comment]
I used the new apis from image fixes.

**Comment by fire** [comment]
This looks good. I think Roughness and Metalness and normal textures are broken and but it's a good start.

**Pull request merged**